libpod/container_api.go

```diff
@@ -453,22 +453,19 @@ func (c *Container) Unmount(force bool) error {
 		}
 	}
 
-	if c.state.State == ContainerStateRunning || c.state.State == ContainerStatePaused {
-		return errors.Wrapf(ErrCtrStateInvalid, "cannot unmount storage for container %s as it is running or paused", c.ID())
-	}
-
-	// Check if we have active exec sessions
-	if len(c.state.ExecSessions) != 0 {
-		return errors.Wrapf(ErrCtrStateInvalid, "container %s has active exec sessions, refusing to unmount", c.ID())
-	}
-
 	if c.state.Mounted {
 		mounted, err := c.runtime.storageService.MountedContainerImage(c.ID())
 		if err != nil {
 			return errors.Wrapf(err, "can't determine how many times %s is mounted, refusing to unmount", c.ID())
 		}
 		if mounted == 1 {
-			return errors.Wrapf(err, "can't unmount %s last mount, it is still in use", c.ID())
+			if c.state.State == ContainerStateRunning || c.state.State == ContainerStatePaused {
```

Comment:
I get moving the c.state.ExecSessions check to here, but I'm not so sure of this state check moving here.  Don't we want to fail out if someone asked to unmount if the container is not in the right state and avoid the mount check?  Seems weird that you'd be able to unmount a container that wasn't running.

Reply:
A good example is described here https://github.com/containers/libpod/issues/1695. There's a counter that will be incremented for each `podman mount` and decremented for each `podman umount`. The current change just makes sure we don't actually perform the unmount when only one "mount" is left (i.e., the one of the running container).

Reply:
Yep, yep, get the only one mount count thing, but still wonder if this one should move.  If you @mehon and others are hip with this move, then OK by me.

Reply:
Looks like this is correct, it will unmount if a container is not mounted and will fail if the container is running and only mounted once.

/lgtm